src/test/java/org/eclipse/yasson/defaultmapping/specific/RecursiveReferenceTest.java

```diff
@@ -0,0 +1,111 @@
+ * The Eclipse Public License is available at http://www.eclipse.org/legal/epl-v10.html
+ * and the Eclipse Distribution License is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *
+ * Contributors:
+ * Jorge Bescos Gascon
+ ******************************************************************************/
+
+package org.eclipse.yasson.defaultmapping.specific;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotEquals;
+import static org.junit.jupiter.api.Assertions.fail;
+
+import java.util.Arrays;
+import java.util.List;
+
+import javax.json.bind.Jsonb;
+import javax.json.bind.JsonbBuilder;
+import javax.json.bind.JsonbConfig;
+import javax.json.bind.JsonbException;
+
+import org.eclipse.yasson.Jsonbs;
+import org.eclipse.yasson.adapters.model.Chain;
+import org.eclipse.yasson.adapters.model.ChainAdapter;
+import org.eclipse.yasson.adapters.model.Foo;
+import org.eclipse.yasson.adapters.model.FooAdapter;
+import org.junit.jupiter.api.Test;
+
+public class RecursiveReferenceTest {
+
+    private static final Jsonb adapterSerializerJsonb = JsonbBuilder.create(new JsonbConfig()
+            .withAdapters(new ChainAdapter(), new FooAdapter()));
+    private final static List<Jsonb> testInstances = Arrays.asList(Jsonbs.defaultJsonb, adapterSerializerJsonb);
+    
+    @Test
+    public void testSerializeRecursiveReference() {
+        Chain recursive = new Chain("test");
+        recursive.setLinksTo(recursive);
+        try {
+            Jsonbs.defaultJsonb.toJson(recursive);
+            fail("Fail is expected");
+        } catch (JsonbException e) {
+            assertEquals(
+                    "Unable to serialize property 'linksTo' from org.eclipse.yasson.adapters.model.Chain",
+                    e.getMessage());
+            assertEquals(
+                    "Recursive reference has been found in class class org.eclipse.yasson.adapters.model.Chain.",
+                    e.getCause().getMessage());
+        }
+    }
+
+    @Test
+    public void testSerializeRepeatedInstance() {
+        testInstances.forEach(jsonb -> checkSerializeRepeatedInstance(jsonb));
+    }
+    
+    private void checkSerializeRepeatedInstance(Jsonb jsonb) {
+        Chain recursive = new Chain("test");
+        recursive.setLinksTo(new Chain("test"));
+        String result = jsonb.toJson(Arrays.asList(recursive, recursive));
+        assertEquals("[{\"linksTo\":{\"name\":\"test\"},\"name\":\"test\"},{\"linksTo\":{\"name\":\"test\"},\"name\":\"test\"}]", result);
+    }
+
+    @Test
+    public void testDeserializeRecursiveReference() {
+        testInstances.forEach(jsonb -> checkDeserializeRecursiveReference(jsonb));
+    }
+    
+    private void checkDeserializeRecursiveReference(Jsonb jsonb) {
+        Chain recursive = jsonb.fromJson("{\"linksTo\":{\"name\":\"test\"},\"name\":\"test\"}", Chain.class);
+        assertNotEquals(recursive.getLinksTo(), recursive);
```

Comment:
I wanted to verify that objects with same values are not the same instance. I know it is tricky that this would happen because there should be some caching. But in case somebody in the future decide to reuse instances because of performance reasons, this test will catch it.

In this case, Chain doesn't implement equals to check values, so it only checks the reference. I have changed the assert to illustrate my intentions:
assertFalse(recursive.getLinksTo() == recursive);

Reply:
Thinking it a bit more, it makes no sense to have this test.

Because 'recursive' and 'recursive.getLinksTo()' are never equals from any point of view. First one has a not null linksTo, and the second has it null.

The test would be valid when if we are testing exactly objects with the same values, in a collection for example:
```
        List<Chain> repeatedObjects = jsonb.fromJson("[{\"name\":\"test\"},{\"name\":\"test\"}]", new TestTypeToken<List<Chain>>(){}.getType());
        assertFalse(repeatedObjects.get(0) == repeatedObjects.get(1));
```
But I think this is out of the scope of this pull request.

I'm going to remove this test.